src/scroll-timeline-base.js

```diff
@@ -510,3 +510,161 @@ export class ScrollTimeline {
     return true;
   }
 }
+
+function getScrollParent(node) {
+  if (node === undefined)
+    return undefined;
+
+  if (node.scrollHeight > node.clientHeight ||
+      node.scrollWidth > node.clientWidth) {
+    return node;
+  } else {
+    return getScrollParent(node.parentNode);
+  }
+}
+
+// https://fantasai.inkedblade.net/style/specs/scroll-animations-1/#view-progress-timelines
+export class ViewTimeline extends ScrollTimeline {
+  // As specced, ViewTimeline has a subject and a source, but
+  // ViewTimelineOptions only has source. Furthermore, there is a strict
+  // relationship between subject and source (source is nearest scrollable
+  // ancestor of subject).
+
+  // Proceeding under the assumption that subject will be added to
+  // ViewTimelineOptions. Inferring the source from the subject if not
+  // explicitly set.
+  constructor(options) {
+    if (options.subject && !options.source) {
+      options.source = getScrollParent(options.subject.parentNode);
+    }
+    super(options);
+
+    const details = scrollTimelineOptions.get(this);
+    details.subject = options && options.subject ? options.subject : undefined;
+    // TODO: what is the default range?
+    details.range = options && options.range ? options.range : 'contain';
+    // TODO: Handle insets.
+  }
+
+  get subject() {
+    return scrollTimelineOptions.get(this).subject;
+  }
+
+  // Proposing that we do away with before and after phase for a timeline and
+  // simply allow timeline time to extend outside the range of 0 to 100%.
+  get phase() {
+    if (!this.subject)
+      return "inactive";
+
+    const container = this.source;
+    if (!container)
+      return "inactive";
+
+    let scrollerStyle = getComputedStyle(container);
+
+    //   if source does not currently have a CSS layout box
+    if (scrollerStyle.display == "none")
+      return "inactive";
+
+    if (container != document.scrollingElement &&
+        (scrollerStyle.overflow == 'visible' ||
+         scrollerStyle.overflow == "clip")) {
+        return "inactive";
+    }
+
+    return "active";
+  }
+
+  // Currently specced as fit with proposal to rename in order to more naturally
+  // support start and end transitions.
+  get range() {
+    return scrollTimelineOptions.get(this).range;
+  }
+
+  get currentTime() {
+    const unresolved = null;
+    if (this.phase === 'inacitve')
+      return unresolved;
+
+    // Compute the offset of the top-left corner of subject relative to
+    // top-left corner of the container.
+    const container = this.source;
+    const target = this.subject;
+
+    const targetBounds = target.getBoundingClientRect();
+    const containerBounds = container.getBoundingClientRect();
```

Comment:
The BoundingClientRect of an overflow scroller is its visual borderbox rect. This means that top will be negative once the user has scrolled down which I think will result in an incorrect startOffset. The boundingClientRect also includes the border, but we should be measuring from the inside of the scroller. Here's a demo: scroll the scroller on https://jsbin.com/zimaxel/edit?html,css,js,output . This means at minimum I think you need to subtract the scrollTop from the containerBounds.

Maybe, however, we should be walking the offsetParent chain to the container and adding up the offsetTop / offsetLeft.

Reply:
Updated to do the walk.  You'll notice a slight tweak to the calculation for viewPos.